controller/src/main/java/io/pravega/controller/store/stream/ZKStream.java

```diff
@@ -111,35 +120,205 @@
+                return String.format(SCOPE_PATH, getScope());
+            } else {
+                return x;
+            }
+        });
+    }
+
+    String getStreamPath() {
+        return streamPath.updateAndGet(x -> {
+            if (Strings.isNullOrEmpty(x)) {
+                return String.format(STREAM_PATH, getScope(), getName());
+            } else {
+                return x;
+            }
+        });
+    }
+
+    private String getRetentionSetPath() {
+        return retentionSetPath.updateAndGet(x -> {
+            if (Strings.isNullOrEmpty(x)) {
+                return String.format(RETENTION_SET_PATH, getScope(), getName());
+            } else {
+                return x;
+            }
+        });
+    }
+
+    private String getRetentionStreamCutRecordPathFormat() {
+        return retentionStreamCutRecordPathFormat.updateAndGet(x -> {
+            if (Strings.isNullOrEmpty(x)) {
+                return String.format(RETENTION_STREAM_CUT_RECORD_PATH, getScope(), getName()) + "/%d";
+            } else {
+                return x;
+            }
+        });    
+    }
+
+    private String getCurrentEpochRecordPath() {
+        return currentEpochRecordPath.updateAndGet(x -> {
+            if (Strings.isNullOrEmpty(x)) {
+                return String.format(CURRENT_EPOCH_RECORD, getScope(), getName());
+            } else {
+                return x;
+            }
+        });
+    }
+
+    private String getEpochRecordPathFormat() {
+        return epochRecordPathFormat.updateAndGet(x -> {
+            if (Strings.isNullOrEmpty(x)) {
+                return String.format(EPOCH_RECORD, getScope(), getName()) + "/%d";
+            } else {
+                return x;
+            }
+        });
+    }
+
+    private String getHistoryTimeSeriesChunkPathFormat() {
+        return historyTimeSeriesChunkPathFormat.updateAndGet(x -> {
+            if (Strings.isNullOrEmpty(x)) {
+                return String.format(HISTORY_TIMESERIES_CHUNK_PATH, getScope(), getName()) + "/%d";
+            } else {
+                return x;
+            }
+        });
+    }
+
+    private String getSegmentSealedEpochPathFormat() {
+        return segmentSealedEpochPathFormat.updateAndGet(x -> {
+            if (Strings.isNullOrEmpty(x)) {
+                return String.format(SEGMENT_SEALED_EPOCH_PATH, getScope(), getName()) + "/%d";
+            } else {
+                return x;
+            }
+        });
+    }
+
+    private String getSegmentsSealedSizeMapShardPathFormat() {
+        return segmentsSealedSizeMapShardPathFormat.updateAndGet(x -> {
+            if (Strings.isNullOrEmpty(x)) {
+                return String.format(SEGMENTS_SEALED_SIZE_MAP_SHARD_PATH, getScope(), getName()) + "/%d";
+            } else {
+                return x;
+            }
+        });
+    }
+
     // region overrides
 
     @Override
     public CompletableFuture<Integer> getNumberOfOngoingTransactions() {
```

Comment:
The logic below this point seems to have little relationship to the previous 200 lines. Can we separate these?

Reply:
the logic below is encapsulated under "region overrides". Is that sufficient or are you suggesting something else?

Reply:
reverted the changes above